### PR TITLE
Make `Startup` a real feature in `App`.

### PR DESCRIPTION
**Pull request by andriyDev**
# Objective

- Make #20318 less likely.

## Solution

- Allow specifying the startup schedule of Apps.
- Run the startup schedule once before looping in all our runners.

Note: This doesn't fully fix the issue of the pink screen, but it does make it MUCH less likely. A proper fix needs either to revert all the embedded asset shader stuffs we did this cycle, OR a much more complex solution that allows rendering to do nothing until its assets load (e.g., a ready single in the render graph or something). This solution is sort of a bandaid.

## Testing

- Ran the `3d_shapes` example and things still seem to be rendering and spawning.

### CODE REVIEW

**Review by IceSentry — APPROVED** [review]
Code LGTM, still needs to update the migration guide and stuff off course.

I confirmed it fixed the pink screen on my m4 macbook

**Review by Zeophlite — APPROVED** [review]
Works on my M1 Pro - no pink flash

**Comment by andriyDev** [comment]
I wanted to mention a possible alternative. We could run RenderStartup in `RenderPlugin::cleanup` instead? This would however have an implicit dependency before `PipelinedRenderingPlugin` since that plugin literally rips out the render app. We'll also miss anything users are doing in cleanup, maybe that's ok though?

**Comment by alice-i-cecile** [comment]
From an ECS perspective, I like making Startup a little less weird / implicit / special, but I dislike building more functionality into subapps.

Long-term I'd like to model `Startup` as a state (allowing it to be rerun and unifying it better), but this wouldn't make that much harder.

I don't have the rendering expertise to judge if it's the best solution there.

**Comment by hymm** [comment]
> We could run RenderStartup in RenderPlugin::cleanup instead? This would however have an implicit dependency before PipelinedRenderingPlugin since that plugin literally rips out the render app. We'll also miss anything users are doing in cleanup, maybe that's ok though?

I think this would be a cleaner solution if we ever move to world's as entities. If we get that we would be able to run RenderStartup as a part of the main world's startup schedule.

**Comment by alice-i-cecile** [comment]
Yep, any "fragmented multiworld", "no render extraction" or "worlds as entities" shaped solution would allow us to nest this more cleanly.

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -152,6 +153,15 @@ impl App {
         }
     }
 
+    /// Runs the default schedules of all sub-apps (starting with the "main" app) once.
+    pub fn startup(&mut self) {
+        if self.is_building_plugins() {
+            panic!("App::startup() was called while a plugin was building.");
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/app.rs:157]: Please leave a comment explaining why this is forbidden.

File: crates/bevy_app/src/main_schedule.rs [hunk 2]
```diff
@@ -55,6 +54,10 @@ use bevy_ecs::{
 #[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash, Default)]
 pub struct Main;
 
+/// The schedule that runs all startup schedules (e.g., [`Startup`]).
```
Comment (alice-i-cecile) [inline crates/bevy_app/src/main_schedule.rs]:
```suggestion
/// The schedule that runs all startup schedules (e.g., [`Startup`], [`PostStartup`]).
```

**Review by tychedelia — APPROVED** [review]
Makes a lot of sense to me (an ECS novice)!

**Review by alice-i-cecile — APPROVED** [review]
I have a few nits, but I think this is fundamentally a fairly reasonable architecture from an ECS perspective. I'm not keen about sub-apps still, or the use of startup rather than strictly relying on states, but I don't think this makes those problems worse to fix in the future.

**Comment by hukasu** [comment]
this would also cause `Startup` to run before the first run of `StateTransitions` right?

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
Actually, I am not okay with changing the behavior of how states work in this PR. This needs a test to confirm that the behavior works as before, and any fixes (which may be none) to ensure that it does so.

**Review by alice-i-cecile — APPROVED** [review]
> StateTransition is a startup schedule, and all the startup schedules got moved, so there will be no change.

- @andriyDev on [Discord](https://discord.com/channels/691052431525675048/749335865876021248/1404955507118047356)

**Comment by alice-i-cecile** [comment]
@cart has requested a look before we merge :)

File: crates/bevy_app/src/app.rs (hunk 1 above)
Comment (andriyDev) [reply crates/bevy_app/src/app.rs:157]: I honestly have no idea why this is forbidden. This was copy-pasted from `fn update` which has the exact same check. I think this is just a sanity check that people aren't calling update while inside of building plugins so they're not doing anything particularly crazy.

**Comment by tychedelia** [comment]
Alternative fix here https://github.com/bevyengine/bevy/pull/20595

**Comment by IceSentry** [comment]
That PR fixes the pink screen issue but I think there's still some ordering weirdness with RenderStartup and we need a PR like this one.

**Comment by alice-i-cecile** [comment]
@tychedelia and I think we should punt on this for now: cutting from the milestone.
